### PR TITLE
staticanalysis/bot: Publish a warning comment when a revision contains files that have zero coverage

### LINKED ISSUE #1823: Add a warning on Phabricator revisions when they are modifying a file with 0 coverage
This could go in the staticanalysis/bot because it uses code coverage data which has already been produced, it doesn't have to parse coverage artifacts and so on like the normal coverage build.
Also, staticanalysis/bot already has all the infrastructure to run stuff when a new Phabricator revision is created and to comment on Phabricator revisions.

### PR DESCRIPTION
**Pull request by marco-c**
Fixes #1823.

### CODE REVIEW

File: src/staticanalysis/bot/tests/conftest.py [hunk 1]
```diff
@@ -441,3 +441,16 @@ def mock_coverity_output():
     Load a real case clang output
     '''
     return os.path.join(MOCK_DIR, 'coverity.json')
+
+
+@pytest.fixture
+@responses.activate
+def mock_coverage():
```
Comment (marco-c) [inline src/staticanalysis/bot/tests/conftest.py]: For some reason this doesn't work and I had to copy its content to the test directly...

File: src/staticanalysis/bot/static_analysis_bot/coverage.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            'publishable': self.is_publishable(),
+        }
+
+
+class Coverage(DefaultAnalyzer):
+    def can_run_before_patch(self):
+        '''
+        Coverage analysis should run only once, after the patch is applied.
+        '''
+        return False
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/coverage.py]: :+1:

File: src/staticanalysis/bot/static_analysis_bot/coverage.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    @stats.api.timed('runtime.coverage')
+    def run(self, revision):
+        '''
+        List all issues found by coverage analysis on specified files
+        '''
+        assert isinstance(revision, Revision)
+
+        # Download zero coverage report.
+        r = requests.get('https://index.taskcluster.net/v1/task/project.releng.services.project.production.code_coverage_bot.latest/artifacts/public/zero_coverage_report.json')  # noqa
+        report = r.json()
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/coverage.py]: It's always nice to have `r.raise_for_status()`

File: src/staticanalysis/bot/static_analysis_bot/coverage.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        zero_coverage_files = set(file_info['name'] for file_info in report['files'] if file_info['uncovered'])
+
+        issues = [
+            CoverageIssue(path, 0, 'This file is uncovered', revision)
+            for path in revision.files
+            if path in zero_coverage_files
+        ]
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/coverage.py]: That's clean and simple, i like it !

File: src/staticanalysis/bot/static_analysis_bot/report/base.py [hunk 5]
```diff
@@ -40,6 +40,10 @@
 You can run this analysis locally with:
 {analyzers}
 '''
+COMMENT_COVERAGE = '''
+In our previous code coverage analysis run, we found some files which had zero coverage and are being modified in this patch:
+{paths}
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/report/base.py]: I would add a call to action : `So please add some unit tests`

File: src/staticanalysis/bot/static_analysis_bot/report/phabricator.py [hunk 6]
```diff
@@ -59,28 +60,42 @@ def publish(self, issues, revision):
...
-        if issues:
+        coverage_issues = [issue for issue in issues if issue.ANALYZER == COVERAGE]
 
+        if issues:
             # First publish inlines as drafts
             inlines = list(filter(None, [
                 self.comment_inline(revision, issue, existing_comments)
                 for issue in issues
-                if issue.ANALYZER != CLANG_FORMAT
+                if issue.ANALYZER not in [CLANG_FORMAT, COVERAGE]
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/report/phabricator.py]: Nit: define a constant `ANALYZERS_WITHOUT_INLINES` with a nice comment

File: src/staticanalysis/bot/tests/conftest.py (hunk 1 above)
Comment (La0) [reply src/staticanalysis/bot/tests/conftest.py]: I had issues with the order of decorators when using pytest + responses

**Pull request merged**